tar.go

```diff
@@ -210,25 +234,43 @@ func untar(tr *tar.Reader, destination string) error {
 			return err
 		}
 
-		if err := untarFile(tr, header, destination); err != nil {
+		if err := untarFile(tr, header, destination, preserveIds); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
 // untarFile untars a single file from tr with header header into destination.
-func untarFile(tr *tar.Reader, header *tar.Header, destination string) error {
+func untarFile(tr *tar.Reader, header *tar.Header, destination string, preserveIds bool) error {
+	var err error
 	switch header.Typeflag {
 	case tar.TypeDir:
-		return mkdir(filepath.Join(destination, header.Name))
+		err = mkdir(filepath.Join(destination, header.Name))
+		break
 	case tar.TypeReg, tar.TypeRegA, tar.TypeChar, tar.TypeBlock, tar.TypeFifo:
-		return writeNewFile(filepath.Join(destination, header.Name), tr, header.FileInfo().Mode())
+		err = writeNewFile(filepath.Join(destination, header.Name), tr, header.FileInfo().Mode())
+		break
 	case tar.TypeSymlink:
-		return writeNewSymbolicLink(filepath.Join(destination, header.Name), header.Linkname)
+		err = writeNewSymbolicLink(filepath.Join(destination, header.Name), header.Linkname)
+		break
 	case tar.TypeLink:
-		return writeNewHardLink(filepath.Join(destination, header.Name), filepath.Join(destination, header.Linkname))
+		err = writeNewHardLink(filepath.Join(destination, header.Name), filepath.Join(destination, header.Linkname))
+		break
 	default:
-		return fmt.Errorf("%s: unknown type flag: %c", header.Name, header.Typeflag)
+		err = fmt.Errorf("%s: unknown type flag: %c", header.Name, header.Typeflag)
+		break
+	}
+	if err != nil {
+		return err
+	}
+
+	if preserveIds && header.Name != "./" && header.Name != "." && runtime.GOOS != "windows" {
```

Comment:
What about if header.Name == "././."?  I'm not sure why you're trying to prevent the os.Lchown() for the current directory.